735 viz dataframe

Jesus89 opened the pull request:
Related to https://github.com/CartoDB/cartoframes/issues/735, https://github.com/CartoDB/cartoframes/issues/746.

This PR adds support for DataFrame visualization. This is the same mechanism we use to get the query from a table, but in this case to obtains a gdf from a df. The geometry is obtained from the available data in the df:
1. Geom column
2. Lat and lng columns

So now you can render a df:

- From local data
```py
import pandas as pd

data = {'latitude': [0, 10, 20, 30], 'longitude': [0, 10, 20, 30]}
df = pd.DataFrame.from_dict(data)

Map(Layer(df, 'width: 30'))
```
- From CARTO

```py
from cartoframes import Dataset

ds = Dataset.from_table('populated_places')

df = ds.download(limit=10)

Map(Layer(df, 'width: 10'))
```

- From Dataset created with a df
```py
import pandas as pd

data = {'latitude': [0, 10, 20, 30], 'longitude': [0, 10, 20, 30]}
df = pd.DataFrame.from_dict(data)

ds = Dataset.from_dataframe(df)

Map(Layer(ds, 'width: 30'))
```

alrocar commented on cartoframes/dataset.py:
```diff
@@ -376,14 +382,81 @@ def get_columns(context, query):
+        return df['longitude']
+    if 'lng' in df:
+        return df['lng']
+    if 'long' in df:
```
maybe we can add `lon` as well

alrocar commented on cartoframes/dataset.py:
```diff
@@ -376,14 +382,81 @@ def get_columns(context, query):
+        return df['wkb_geometry']
+    if 'geom' in df:
+        return df['geom']
+    if 'geojson' in df:
```
About geojson, if we support them, we should decode them properly right? I guess decode_geom does not support geojson.

Jesus89 replied:
I supposed that this was an alias for `geometry` (https://github.com/CartoDB/cartodb/blob/master/services/importer/lib/importer/georeferencer.rb#L14).
IMO we can remove this case since we have good support for geojson in CF

alrocar replied:
let's remove it then.

Jesus89 commented:
Yes, I have in mind to add tests. I'm thinking now if we could move the `get_query` and `get_geodataframe' utilities to another file (?) The Dataset file is pretty big now.

alrocar commented:
>  I'm thinking now if we could move the get_query and `get_geodataframe' utilities to another file

LGTM

andy-esch commented:
👀

andy-esch commented on cartoframes/dataset.py:
```diff
@@ -376,14 +382,85 @@ def get_columns(context, query):
+            if lat_column is not None and lng_column is not None:
+                df['geometry'] = _compute_geometry_from_latlng(lat_column, lng_column)
+            else:
+                raise ValueError('DataFrame has no geographic data.')
```
How about something like this:
"No geographic data found. If a geometry exists, change the column name to `the_geom` or `geom` or ensure it is a GeoDataFrame with a valid geometry. If there are latitude/longitude columns, rename to `lat`/`lng`."

This way people have a route forward if they do indeed have geographic information.

andy-esch replied:
Or better yet we can give them the full list of valid names/formats.

Jesus89 replied:
What about 

> No geographic data found. If a geometry exists, change the column name (geometry, the_geom, wkt_geometry, wkb_geometry, geom, wkt, wkb) or ensure it is a DataFrame with a valid geometry. If there are latitude/longitude columns, rename to (latitude, lat), (longitude, lng, lon, long).

andy-esch replied:
👍

andy-esch commented on cartoframes/viz/layer.py:
```diff
@@ -104,7 +112,9 @@ def __init__(self,
     """Set a Source class from the input"""
-    if isinstance(source, (str, list, dict, Dataset)):
+    if isinstance(source, (str, list, dict, Dataset)) or \
+       isinstance(source, pandas.DataFrame) or \
```
Why not include `pandas.DataFrame` in the tuple above with the other types?

Jesus89 replied:
True that

Jesus89 commented:
Hey, I have merged `develop` with the sharing viz changes, updated the `data` namespace, and added tests to cover the geometry generation. So it's ready for review :smile:

oleurud commented on cartoframes/data/dataset.py:
```diff
@@ -1,15 +1,15 @@
-from carto.exceptions import CartoException, CartoRateLimitException
+from carto.exceptions import CartoException
 
+from .utils import decode_geometry, compute_query, compute_geodataframe, get_columns, \
```
The max length is 120. You can put the constant in the same line

Jesus89 replied:
I'm not used to 120, but I can change it

oleurud commented on cartoframes/data/utils.py:
```diff
@@ -0,0 +1,150 @@
+
+
+def compute_geodataframe(dataset):
+    if dataset._df is not None:
```
```suggestion
    if dataset.dataframe is not None:
```

Jesus89 replied:
Oh, I didn't notice this getter :+1:

oleurud commented on cartoframes/data/utils.py:
```diff
@@ -0,0 +1,150 @@
+from ..columns import Column
+
+try:
+    import geopandas
```
Why do we need that? Why is not a new dependency?

Jesus89 replied:
This is not a dependency. CARTOframes works without geopandas, but this library is required for some situations. This comes from the past versions, but you can create a ticket to discuss this in future versions of CF.

andy-esch replied:
geopandas is a heavy dependency, so I like the idea of making it optional since it's only required for a couple of features. This is a common pattern. E.g., matplotlib is not required for pandas but when installed `dataframe.plot()` returns plots.

oleurud commented on cartoframes/data/utils.py:
```diff
@@ -0,0 +1,150 @@
+                                     ', '.join(LAT_COLUMN_NAMES),
+                                     ', '.join(LNG_COLUMN_NAMES)
+                                 ))
+        return geopandas.GeoDataFrame(df)
```
It will fail without geopandas dependency

Jesus89 replied:
Yep, we need to check the flag here. Thanks

oleurud commented on cartoframes/data/utils.py:
```diff
@@ -0,0 +1,150 @@
+
+def compute_geodataframe(dataset):
+    if dataset._df is not None:
+        df = dataset._df.copy()
```
```suggestion
        df = dataset.dataframe.copy()
```

oleurud replied:
I don't like to copy it. It could be a performance problem (in memory). I am thinking about it.

Jesus89 replied:
If we don't copy the df to generate the gdf, the original df will be modified with an extra column. I'm not sure if we want this or not. Maybe it's a good thing to have the geometry column already there.

andy-esch replied:
Yeah, I worry about performance too. E.g., if a user has a dataframe that's 1GB, which isn't uncommon if they have complex polygons.

andy-esch replied:
What about a warning (or logging.info) that a new column is being added to the original dataframe?

Jesus89 replied:
We have removed the copy() by now. But I think we should revisit it in the future.

Regarding the logging info, this will be displayed always when there is a df visualization, but if you consider that it is OK to have a warning we can add it

oleurud commented on cartoframes/viz/kuviz.py:
```diff
@@ -20,7 +20,7 @@ def __init__(self, id, url, name, privacy=PRIVACY_PUBLIC):
     @classmethod
     def create(cls, html, name, context=None, password=None):
-        from cartoframes.auth import _default_context
+        from ..auth import _default_context
```
Why do you prefer a relative path?

Jesus89 replied:
I think it is a common practice to use relative imports in the project and absolute imports for your dependencies. It requires usually fewer chars and it decouples your implementation from the public API.

andy-esch replied:
Agree, I like relative imports for project module imports

oleurud commented on cartoframes/viz/layer.py:
```diff
@@ -1,17 +1,25 @@
 from .legend import Legend
+from ..data import Dataset
+
+try:
```
Same comment as Dataset one

oleurud commented on cartoframes/data/utils.py:
```diff
@@ -0,0 +1,150 @@
+import time
```
I really like this file, making the rest cleaner

Jesus89 replied:
:heart:

andy-esch commented on cartoframes/data/dataset.py:
```diff
@@ -515,49 +497,5 @@ def _get_geom_col_type(df):
-                            pass
-    return None
+def _first_not_null_value(array):
+    return array.loc[~array.isnull()].iloc[0]
```
What happens if they're all null? I think there will be an `IndexError`

Jesus89 replied:
True that. We can rename the method and use an if everywhere it is used. cc @alrocar

alrocar commented on cartoframes/data/dataset.py:
```diff
@@ -504,7 +493,7 @@ def _get_geom_col_type(df):
     try:
-        geom = _decode_geom(_first_not_null_value(df, geom_col))
+        geom = decode_geometry(_first_value(df[geom_col]))
     except IndexError:
```
Will this method raise an IndexError in any case? Maybe we could add a test case ;)

Jesus89 replied:
This Exception won't be raised anymore. What case do you want to cover with tests?

alrocar replied:
If first value returns `None` I guess `decode_geometry` will raise an exception, right? (not an IndexError) should we cover that case with a test?

Jesus89 replied:
decode_geometry has an if internally, so the None case is ignored

alrocar replied:
🤦‍♂ I missread the code, you are right!
